### ISSUE TITLE
Caching Resource Tokens

### ISSUE THREAD

**Issue by danutzplusplus** [Azure/azure-cosmosdb-dotnet#635]
Hello,

I've two questions about cosmos db resource tokens (or better yet, permissions in my case). and I'd really appreciate if someone with more experience could give some feedback. Thanks.

Considering that resource tokens / permissions have a lifetime I'd like to cache them on the client. 
I imagine there's nothing wrong with that?

And if that is ok, I'm trying to understand if the simple call to DocumentClient.ReadPermissionFeedAsync(user.SelfLink) causes for a new token to be issued to the client? I imagine the DocumentClient.CreatePermissionAsync(user.SelfLink, permission) just grants the permission, and it's the ReadPermission that requests the token for actual use?

I've examined the resource token, and can see that the signature changes with every new token, but I imagine that is because of the datetime value that's part of the token message. Does anyone know if that datetime is the actual value used to determine if the token is expired? (on a side now, how exactly does the cosmos server validate the HMAC if it includes a datetime? do they both aproximate to a value or how? this leaves a bit bewildered)

Basically, I'd like to find a way to programmatically know how much life my resource token has, in order to read a new one. 

Is any of this possible, or am I under some misinformation?

**Comment by danutzplusplus** [comment]
Ok, it seems I've found an answer for the first half of my question: 
[resource tokens](https://docs.microsoft.com/en-us/rest/api/cosmos-db/access-control-on-cosmosdb-resources?redirectedfrom=MSDN#resource-tokens)

> When a permission resource is acted upon on (POST, GET, PUT), a new resource token is generated. 

Which leaves the second part. How can I programmatically calculate the lifetime of the token? And is that even possible? I see no explicit lifetime information on the token. Or am I missing something? Thanks

**Comment by danutzplusplus** [comment]
Ok, seems there's a timestamp on the permission itself. [permission](https://docs.microsoft.com/en-us/rest/api/cosmos-db/permissions)

> _ts | It is a system generated property. It specifies the last updated timestamp of the resource. The value is a timestamp.

I wonder if that can be used

**Comment by briansboyd** [comment]
Hi @danutzplusplus,
You can specify the lifetime of each Resource Token and then easily calculate the expiry time.
What I do is then create a **Tuple<string, DateTime>** that is passed around and/or cached that includes both the resource token and this calculate expiry time.
Cheers,
Brian
